File: view/src/com/bstek/dorado/view/widget/base/ProgressBar.java
/*
 * This file is part of Dorado 7.x (http://dorado7.bsdn.org).
 * 
 * Copyright (c) 2002-2012 BSTEK Corp. All rights reserved.
 * 
 * This file is dual-licensed under the AGPLv3 (http://www.gnu.org/licenses/agpl-3.0.html) 
 * and BSDN commercial (http://www.bsdn.org/licenses) licenses.
 * 
 * If you are unsure which license is appropriate for your use, please contact the sales department
 * at http://www.bstek.com/contact.
 */

package com.bstek.dorado.view.widget.base;

import com.bstek.dorado.annotation.ClientObject;
import com.bstek.dorado.annotation.ClientProperty;
import com.bstek.dorado.annotation.XmlNode;
import com.bstek.dorado.common.ClientType;
import com.bstek.dorado.view.annotation.Widget;
import com.bstek.dorado.view.widget.Control;

/**
 * @author Benny Bao (mailto:benny.bao@bstek.com)
 * @since 2010-8-9
 */
@Widget(name = "ProgressBar", category = "General",
		dependsPackage = "base-widget")
@ClientObject(prototype = "dorado.widget.ProgressBar",
		shortTypeName = "ProgressBar")
@XmlNode(clientTypes = { ClientType.DESKTOP, ClientType.TOUCH })
public class ProgressBar extends Control {
	private float minValue;
	private float maxValue = 100;
	private String value;
	private String textPattern = "{percent}%";
	private boolean showText = true;
	private boolean effectEnable = false;

	public float getMinValue() {
		return minValue;
	}

	public void setMinValue(float minValue) {
		this.minValue = minValue;
	}

	@ClientProperty(escapeValue = "100")
	public float getMaxValue() {
		return maxValue;
	}

	public void setMaxValue(float maxValue) {
		this.maxValue = maxValue;
	}

	public String getValue() {
		return value;
	}

	public void setValue(String value) {
		this.value = value;
	}

	@ClientProperty(escapeValue = "{percent}%")
	public String getTextPattern() {
		return textPattern;
	}

	public void setTextPattern(String textPattern) {
		this.textPattern = textPattern;
	}

	@ClientProperty(escapeValue = "true")
	public boolean isShowText() {
		return showText;
	}

	public void setShowText(boolean showText) {
		this.showText = showText;
	}

	public boolean isEffectEnable() {
		return effectEnable;
	}

	public void setEffectEnable(boolean effectEnable) {
		this.effectEnable = effectEnable;
	}

}

File: client/client/scripts/dorado/widget/base/progress-bar.js
/*
 * This file is part of Dorado 7.x (http://dorado7.bsdn.org).
 * 
 * Copyright (c) 2002-2012 BSTEK Corp. All rights reserved.
 * 
 * This file is dual-licensed under the AGPLv3 (http://www.gnu.org/licenses/agpl-3.0.html) 
 * and BSDN commercial (http://www.bsdn.org/licenses) licenses.
 * 
 * If you are unsure which license is appropriate for your use, please contact the sales department
 * at http://www.bstek.com/contact.
 */

/**
 * @component Base
 * @class 进度条
 * @extends dorado.widget.Control
 */
dorado.widget.ProgressBar = $extend(dorado.widget.Control, /** @scope dorado.widget.ProgressBar.prototype */ {
	$className: "dorado.widget.ProgressBar",
	selectable: false,
	
	ATTRIBUTES: /** @scope dorado.widget.ProgressBar.prototype */ {
		className: {
			defaultValue: "d-progress-bar"
		},
		height: {
			independent: true
		},
        /**
         * 该进度条显示范围的最小值，默认为0。
         * @attribute
         * @default 0
         * @type int|float
         */
		minValue: {
			defaultValue: 0
		},
        /**
         * 该进度条显示范围的最大值，默认为100。
         * @attribute
         * @default 100
         * @type int|float
         */
		maxValue: {
			defaultValue: 100
		},
        /**
         * 该进度条是否显示文字，默认为true。
         * @attribute
         * @default true
         * @type boolean
         */
		showText: {
			defaultValue: true
		},
        /**
         * 进度条当前值
         * @attribute
         * @type int|float
         */
		value: {},
        /**
         * 该进度条限制文本的格式，进度使用{percent}来替换，默认值为{percent}%。
         * @attribute
         * @default "{percent}%"
         * @type String
         */
		textPattern: {
			defaultValue: "{percent}%"
		}
	},

	createDom: function() {
		var bar = this, doms = {}, dom = $DomUtils.xCreate({
			tagName: "div",
			className: bar._className,
			content: [
				{
					tagName: "span",
					className: "msg",
					contextKey: "msg"
				},
				{
					tagName: "span",
					className: "bar",
					contextKey: "bar",
					content: {
						tagName: "span",
						className: "bar-msg",
						contextKey: "barMsg"
					}
				}
			]
		}, null, doms);

		bar._doms = doms;

		return dom;
	},
    
	refreshDom: function(dom) {
		$invokeSuper.call(this, arguments);
		var bar = this, min = bar._minValue, max = bar._maxValue, value = bar._value || 0, doms = bar._doms,
			percent = value / (max - min), showText = bar._showText, pattern = bar._textPattern || "";

		if (percent == 0) {
			$fly(dom).addClass("d-rendering");
		}
		
		if (percent >= 0 && percent <= 1) {
			$fly(doms.bar).css("width", percent * 100 + "%");
		}

		var $msg = $fly([doms.msg, doms.barMsg]).css("width", dom.offsetWidth);
		if (showText) {
			$msg.text(pattern.replace("{percent}", parseInt(percent * 100, 10)));
		} else {
			$msg.empty();
        }
		
		if (percent == 0) {
			$fly(dom).removeClass("d-rendering");
		}
	}
});
